src/rpc/mining.cpp

```diff
@@ -933,16 +944,376 @@ static UniValue estimaterawfee(const JSONRPCRequest& request)
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Bogus crap sketch.");
+        } else if (status == READ_STATUS_FAILED) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Failed to complete block though all transactions were apparently found. Could be random short ID collision; requires full block instead.");
+        } else if (status == READ_STATUS_CHECKBLOCK_FAILED) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Checkblock failed.");
+        }
+        CDataStream ssBlock(SER_NETWORK, PROTOCOL_VERSION);
+        ssBlock << *pblock;
+
+        ret.pushKV("blockhex", HexStr(ssBlock.begin(), ssBlock.end()));
+    } else {
+        ret.pushKV("block_tx_req", HexStr(ssReq.begin(), ssReq.end()));
+        ret.pushKV("found_transactions", HexStr(ssFound.begin(), ssFound.end()));
+    }
+    return ret;
+}
+
+UniValue consumegetblocktxn(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 2)
+        throw std::runtime_error(
+            "consumegetblocktxn full_block block_tx_req\n"
+            "Consumes a transaction request for a compact block sketch."
+            "Arguments:\n"
+            "1. \"full_block\" (string, required), Hex serialied block that corresponds to the block request `block_tx_req`.\n"
+            "2. \"block_tx_req\" (string, required), Hex serialied BlockTransactionsRequest, aka getblocktxn network message.\n"
+            "\nResult\n"
+            "block_transactions  (hex) The serialized list of found transactions aka BlockTransactions\n"
+            "\nExamples:\n"
+            + HelpExampleCli("consumegetblocktxn", "<block_tx_req>")
+        );
+
+
+    CBlock block;
+    std::vector<unsigned char> block_bytes(ParseHex(request.params[0].get_str()));
+    CDataStream ssBlock(block_bytes, SER_NETWORK, PROTOCOL_VERSION);
+    ssBlock >> block;
+
+    // Take in BlockTransactionsRequest, return BlockTransactions
+    std::vector<unsigned char> block_req(ParseHex(request.params[1].get_str()));
+    CDataStream ssReq(block_req, SER_NETWORK, PROTOCOL_VERSION);
+
+    BlockTransactionsRequest req;
+    ssReq >> req;
+
+    BlockTransactions resp(req);
+    for (size_t i = 0; i < req.indexes.size(); i++) {
+        if (req.indexes[i] >= block.vtx.size()) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Peer sent us a getblocktxn with out-of-bounds tx indices");
+        }
+        resp.txn[i] = block.vtx[req.indexes[i]];
+    }
+
+    CDataStream ssResp(SER_NETWORK, PROTOCOL_VERSION);
+    ssResp << resp;
+
+    return HexStr(ssResp.begin(), ssResp.end());
+}
+
+UniValue finalizecompactblock(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 3)
+        throw std::runtime_error(
+            "finalizecompactblock compact_hex block_transactions found_transactions\n"
+            "Takes the two transaction lists, fills out the compact block and attempts to validate it."
```

Comment:
"validate" here means checking if the block is complete, right? I don't see it being submitted for inclusion in the blockchain.

Reply:
will elaborate

Reply:
changed to "finalize"